samza-core/src/main/java/org/apache/samza/table/remote/RemoteTableDescriptor.java

```diff
@@ -0,0 +1,153 @@
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.apache.samza.table.remote;
+
+import java.io.ByteArrayOutputStream;
+import java.io.IOException;
+import java.io.ObjectOutputStream;
+import java.util.Base64;
+import java.util.HashMap;
+import java.util.Map;
+
+import org.apache.samza.SamzaException;
+import org.apache.samza.operators.BaseTableDescriptor;
+import org.apache.samza.table.TableSpec;
+import org.apache.samza.util.RateLimiter;
+
+import com.google.common.base.Preconditions;
+
+
+/**
+ * Table descriptor for remote store backed tables
+ *
+ * @param <K> the type of the key
+ * @param <V> the type of the value
+ */
+public class RemoteTableDescriptor<K, V> extends BaseTableDescriptor<K, V, RemoteTableDescriptor<K, V>> {
+  // Input support for a specific remote store (required)
+  private TableReadFunction<K, V> readFn;
+
+  // Output support for a specific remote store (optional)
+  private TableWriteFunction<K, V> writeFn;
+
+  private RateLimiter readRateLimiter;
+
+  private RateLimiter writeRateLimiter;
+
+  /**
+   * Construct a table descriptor instance
+   * @param tableId Id of the table
+   */
+  public RemoteTableDescriptor(String tableId) {
+    super(tableId);
+  }
+
+  @Override
+  public TableSpec getTableSpec() {
+    validate();
+
+    Map<String, String> tableSpecConfig = new HashMap<>();
+    generateTableSpecConfig(tableSpecConfig);
+
+    // Serialize and store reader/writer functions with config
+    tableSpecConfig.put(RemoteReadableTable.READ_FN, serializeObject("read function", readFn));
+
+    if (writeFn != null) {
+      tableSpecConfig.put(RemoteReadableTable.WRITE_FN, serializeObject("write function", writeFn));
+    }
+
+    // Serialize and store reader/writer rate limiters with config
+    if (readRateLimiter != null) {
+      tableSpecConfig.put(RemoteReadableTable.READ_RLIMIT, serializeObject("read rate limiter", readRateLimiter));
+    }
+
+    if (writeRateLimiter != null) {
+      tableSpecConfig.put(RemoteReadableTable.WRITE_RLIMIT, serializeObject("write rate limiter", writeRateLimiter));
+    }
+
+    return new TableSpec(tableId, serde, RemoteTableProviderFactory.class.getName(), tableSpecConfig);
+  }
+
+  /**
+   * Use specified TableReadFunction with remote table.
+   * @param readFn read function instance
+   * @return this table descriptor instance
+   */
+  public RemoteTableDescriptor<K, V> withReadFunction(TableReadFunction<K, V> readFn) {
+    Preconditions.checkNotNull(readFn, "null read function");
+    this.readFn = readFn;
+    return this;
+  }
+
+  /**
+   * Use specified TableWriteFunction with remote table.
+   * @param writeFn write function instance
+   * @return this table descriptor instance
+   */
+  public RemoteTableDescriptor<K, V> withWriteFunction(TableWriteFunction<K, V> writeFn) {
```

Comment:
Can we add 2 more convenient methods: 1 for read and 1 for write, which only take numbers, internally we would create the rate limiter instance.

Reply:
just added them.